cpp/src/utilities/device_operators.cuh

```diff
@@ -24,103 +24,130 @@
  * ---------------------------------------------------------------------------**/
 
 #include <cudf/cudf.h>
-#include <utilities/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
-#include <cudf/utilities/legacy/wrapper_types.hpp>
+#include <utilities/cudf_utils.h>  // need for CUDA_HOST_DEVICE_CALLABLE
 #include <cudf/utilities/error.hpp>
+#include <cudf/utilities/legacy/wrapper_types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <type_traits>
 
 namespace cudf {
 
 // ------------------------------------------------------------------------
 // Binary operators
 /* @brief binary `sum` operator */
 struct DeviceSum {
-    template<typename T>
-    CUDA_HOST_DEVICE_CALLABLE
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs + rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
+  template <typename T,
+            typename std::enable_if_t<cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return T{DeviceSum{}(lhs.time_since_epoch(), rhs.time_since_epoch())};
+  }
+
+  template <typename T,
+            typename std::enable_if_t<!cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return lhs + rhs;
+  }
+
+  template <typename T>
+  static constexpr T identity() {
+    return T{0};
+  }
 };
 
 /* @brief `count` operator - used in rolling windows */
 struct DeviceCount {
-    template<typename T>
-    CUDA_HOST_DEVICE_CALLABLE
-    T operator() (const T &, const T &rhs) {
-        return rhs + T{1};
-    }
-
-    template<typename T>
-    static constexpr T identity() { return T{0}; }
+  template <typename T,
+            typename std::enable_if_t<cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return T{DeviceCount{}(lhs.time_since_epoch(), rhs.time_since_epoch())};
+  }
+
+  template <typename T,
+            typename std::enable_if_t<!cudf::is_timestamp<T>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T&, const T& rhs) {
+    return rhs + T{1};
+  }
+
+  template <typename T>
+  static constexpr T identity() {
+    return T{0};
+  }
 };
 
 /* @brief binary `min` operator */
-struct DeviceMin{
-    template<typename T>
-    CUDA_HOST_DEVICE_CALLABLE
-    T operator() (const T &lhs, const T &rhs) {
-        return lhs <= rhs? lhs: rhs;
-    }
-
-    template<typename T>
-    static constexpr T identity() { return std::numeric_limits<T>::max(); }
+struct DeviceMin {
+  template <typename T>
+  CUDA_HOST_DEVICE_CALLABLE T operator()(const T& lhs, const T& rhs) {
+    return lhs <= rhs ? lhs : rhs;
```

Comment:
We should try to use `min` rather than the ternary operator here for types that support it. NVCC seems to compile min to a single min instruction, whereas it compiles ternary to 2 instructions: 
https://godbolt.org/z/QEthM_

Reply:
I realized I had `int16_t` in that code. If you change to `int`, the compiler generates `min` for either expression.  However, it will not generate `min()` from the ternary expression if the type is anything but `int32_t` or `int64_t`. Interestingly, not even `float` or `double`.  But `min` does work for int8-64 and float32-64, so I think we should use min/max if possible.